### PR TITLE
Add Provider and Course models

### PR DESCRIPTION
**Pull request by duncanjbrown**
### Context

As we prepare to take data from Find, we need a place to put it. The first step is to create the models and drive e2e tests with them, so that we know what data we want and how we want it.

We are creating two new concepts in our system: `Provider` and `Course`. This commit introduces those models and refactors our code around them.

**The concept of `Site` is also significant but is not included in this PR** because pre-existing code doesn't consider it and we want to keep the change small. It will be added in a future PR.

The next step once this is merged is to drive the simpler of the use cases from the ADR #318:

- applications being associated with courses/sites
- providers seeing their courses in the UI

Both of these changes can be driven from the Provider Interface side by beefing up `provider_interface/see_applications_spec.rb`.

### Changes proposed in this pull request

- Create `Provider` and `Course` models
- Remove redundant `*_ucas_code` fields from `ApplicationChoice`
- Build associations between `Provider`s, `Course`s and `ApplicationChoice`s
- Fix the broken specs

### Guidance to review

Look at `docs/domain-model.png`.

Are the fields correct? Do we have indexes in the right places? Is `schema.rb` ok?

Silly mistakes in models and associations.

### Link to Trello card

https://trello.com/c/hmJQSExM/1099-model-courses-providers-and-populate-with-find-data-%F0%9F%8F%88-needs-kickoff

⭐️ Includes first commits from @stevehook

### CODE REVIEW

**Comment by duncanjbrown** [comment]
Re ❌ on the build, I think DockerHub was briefly down yesterday

Screenshot 2019-10-16 at 09 30 33

File: app/controllers/provider_interface/application_choices_controller.rb [hunk 1]
```diff
@@ -13,7 +12,7 @@ def index
     def show
       application_choice = ApplicationChoice
         .includes(:application_form)
-        .where(provider_ucas_code: current_user.provider_ucas_code)
+        .for_provider_code(current_user.provider.code)
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/controllers/provider_interface/application_choices_controller.rb]: It's only the struct for now — can revisit when we implement `ProviderUser`?

[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/controllers/provider_interface/application_choices_controller.rb]: :+1: will rename the scope

File: app/services/provider_interface/application_choice_presenter.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ def full_name
     end
 
     def course_name
-      "Course #{application_choice.course_ucas_code}"
+      "Course #{application_choice.course.code}"
```
[the remark this answers is not in the document]
Comment (duncanjbrown) [reply app/services/provider_interface/application_choice_presenter.rb]: 😬 specs should have caught this

**Comment by duncanjbrown** [comment]
@tijmenb feedback should be addressed

**Comment by duncanjbrown** [comment]
⚠️ **DO NOT MERGE UNTIL AFTER SHOW AND TELL** ⚠️

**Comment by duncanjbrown** [comment]
Rebasing

File: app/models/candidate.rb [hunk 3]
```diff
@@ -13,9 +13,13 @@ class Candidate < ApplicationRecord
   has_many :application_forms
 
   def current_application
-    application_form = application_forms.first_or_create
-    # TODO: this is a temporary thing until candidates can choose their course
-    application_form.application_choices.first_or_create(provider_ucas_code: 'ABC')
-    application_form
+    choice = ApplicationChoice.first_or_create do |ac|
```
Comment (tvararu) [inline app/models/candidate.rb]: Oops, this fetches the first `ApplicationChoice` (which could belong to a different user, and be completed etc) and then reassigns it further down to belong to the current user's application form.

So:

```ruby
[1] pry(main)> ApplicationForm.all.map(&:destroy!)
[2] pry(main)> ApplicationChoice.all.map(&:destroy!)
[3] pry(main)> ApplicationForm.count
=> 0
[4] pry(main)> ApplicationChoice.count
=> 0
[5] pry(main)> first_user = Candidate.first
[6] pry(main)> second_user = Candidate.second

[7] pry(main)> first_user.current_application
[8] pry(main)> ApplicationForm.count
=> 1 # One more
[9] pry(main)> ApplicationChoice.count
=> 1 # One more
[10] pry(main)> second_user.current_application
[11] pry(main)> ApplicationForm.count
=> 2 # One more
[12] pry(main)> ApplicationChoice.count
=> 1 # Oops!!!
[13] pry(main)> ApplicationChoice.first.application_form.candidate_id == second_user.id
=> true # Oops!!!
```

Comment (stevehook) [reply app/models/candidate.rb]: Thanks @tvararu - will get a fix for that raised asap

**Comment by tvararu** [comment]
@stevehook I’ve submitted a PR for it earlier, it should be addressed :+1:

**Comment by stevehook** [comment]
thanks @tvararu, just saw that - got a bit confused for a moment when I pulled master and looked for that broken code!

**Pull request merged**